### ISSUE TITLE
Moving to a UI Library & Client-Only-Architecture

### ISSUE THREAD

**Issue by vincentscode** [struandw/kobold-training-club#129]
Hi, I have seen the React-Rewrite-Branch and while working on this repository some time ago I was wondering why there even is a client and a server when everything that is currently done in the python backend could be done just as well in the browser, as there is nothing security-critical about it or anything like that.

Given the intent to move towards a UI Library I wanted to give my input on this and suggest maybe using Vue.js for a less heavy and easier to learn frontend and moving away from a backend.
This would also allow hosting the page for free with GitHub pages and by that also resolve issues like #117.
#14 and alike would also likely be simplified by using a unified approach.

Overall the project seems a little deserted with no changes in the past 5 months, maybe this could also be seen as a restart in that regard.

I look forward to hearing your opinion on this! :)

**Comment by struandw** [comment]
Hi Vincent,

Thanks for submitting such a detailed issue!

The primary reason for there being a frontend-backend architecture is that there will always be a need for a database - to wit, a centralised repository of monsters.

However, there’s a good chance that this could be migrated to a single DB instance hosted in AWS or Firebase with queries being assembled in the front-end. I’ll look at this in the next major architecture session.

While I appreciate the suggestion of Vue.js I’m actually pretty comfortable working in React. It’s not perfect, but I think it’ll probably be a fairly good fit for rewriting the site, which I definitely intend on doing. Hosting on GitHub Pages is a particularly good idea, though - thank you for the suggestion!

Regarding #14 and the like, I intend on building a more reactive front-end layout in Bootstrap. While the current app is written in Bootstrap, it was my first time doing web stuff in several years, and I made some pretty bad design choices early on, which I fully intend on remedying.

As far as the project seeming deserted, you have my sincere apologies for that. I’m a freelance developer and the amount of work I’ve had in the least half-year or so means I’ve had little time to devote to side projects! However, I think I have a clear patch coming up over the summer, so I should be able to devote some time to it then!

TL;DR: I want to keep a DB, but assemble queries in front-end, and will probably use Firebase or GH Pages to host. I’ve been busy but will be starting work on this again soon!

**Comment by iykdw** [comment]
Hi Vincent,

Thanks for submitting such a detailed issue!

The primary reason for there being a frontend-backend architecture is that there will always be a need for a database - to wit, a centralised repository of monsters.

However, there’s a good chance that this could be migrated to a single DB instance hosted in AWS or Firebase with queries assembled in the front-end . I’ll look at this in the next major architecture session.

While I appreciate the suggestion of Vue.js I’m actually pretty comfortable working in React. It’s not perfect, but I think it’ll probably be a fairly good fit for rewriting the site, which I definitely intend on doing. Hosting on GitHub Pages is a particularly good idea, though - thank you for the suggestion!

Regarding #14 and the like, I intend on building a more reactive front-end layout in Bootstrap. While the current app is written in Bootstrap, it was my first time doing web stuff in several years, and I made some pretty bad design choices early on, which I fully intend on remedying.

As far as the project seeming deserted, you have my sincere apologies for that. I’m a freelance developer and the amount of work I’ve had in the least half-year or so means I’ve had little time to devote to side projects! However, I think I have a clear patch coming up over the summer, so I should be able to devote some time to it then!

TL;DR: I want to keep a DB, but assemble queries in front-end, and will probably use Firebase or GH Pages to host. I’ve been busy but will be starting work on this again soon!

**Comment by vincentscode** [comment]
Hi Struan,

Thanks for your response!

Having it laid out like this a database structure indeed seems useful as a single "source of truth".
The Firebase free-tier seems like like a great for it considering it's great API and easy way of structuring data like this.

React is of course also a great choice, I was just held back from contributing to the core of the project by constructs like this scattered around the code and think that using any UI library will greatly improve readability as well as make contributing way easier. :)
https://github.com/struandw/kobold-training-club/blob/8e5c1813028de6d81b2d4113d0e14cd02b047ed8/ktc/static/js/encounter-manager.js#L91

Bootstrap is also a good choice considering it's one of the most-common component-libraries. The only reason to consider using alternatives such as Tailwind might be how huge and bloated Bootstrap has gotten over the years. Considering Bootstraps great integration with React may very well make up for that though.

I'd be very happy to help with the rewrite or the migration to a database in the cloud, so feel free to notify me if the chance arises! I myself am a freelance developer so I can totally understand the lack of free-time for side-projects and I hope I did not come across as demanding. :)

I'm looking forward to the future of this project!
